File: lib/features/auth/auth_screen.dart
import 'dart:convert';
import 'dart:math' show Random;
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:google_sign_in/google_sign_in.dart';
import 'package:go_router/go_router.dart';
import 'package:http/http.dart' as http;
import 'package:package_info_plus/package_info_plus.dart';
import '../../core/config/app_config.dart';
import '../../core/services/migration_service.dart';
import '../../core/services/tag_service.dart';
import '../../core/services/user_service.dart';
import '../../core/theme/app_theme.dart';

/// Photos d'illustration de l'en-tête (voyage / naissance / famille — thèmes
/// de l'app), une tirée au hasard à chaque ouverture de cet écran (David
/// 22.09.26 : "une photo de ta bibliothèque et ça tourne"). Avant la
/// connexion il n'y a par définition aucun souvenir à montrer (contrairement
/// au dashboard, voir home_screen.dart::_maybePickHero) — banque Unsplash,
/// licence gratuite, aucune attribution requise, URLs vérifiées le 22.09.26
/// (chaque `images.unsplash.com/photo-…` répond 200, image/jpeg).
const _authHeaderPhotos = <String>[
  'https://images.unsplash.com/photo-1543342384-1f1350e27861', // parents + bébé
  'https://images.unsplash.com/photo-1591161555818-7b9debeccc07', // bébé, noir et blanc
  'https://images.unsplash.com/photo-1610901158532-d246c011729e', // enfant, bonnet
  'https://images.unsplash.com/photo-1637184572364-a231e8b4c716', // parents portant bébé
  'https://images.unsplash.com/photo-1581952975975-08cd95a728d4', // couple, mariage
  'https://images.unsplash.com/photo-1562494794-d59c886e2bf2', // famille, mains
  'https://images.unsplash.com/photo-1636830632657-1dcda9360a3a', // parents portant bébé
  'https://images.unsplash.com/photo-1475503572774-15a45e5d60b9', // famille, plage
  'https://images.unsplash.com/photo-1628705250580-80b96d4657f6', // famille, cerf-volant
  'https://images.unsplash.com/photo-1539635278303-d4002c07eae3', // famille, plein air
];

/// Écran de connexion / inscription — le formulaire reste volontairement
/// simple (juste email + mot de passe et Google), mais porte désormais en
/// en-tête une photo illustrative (voir _authHeaderPhotos) plutôt que rien.
class AuthScreen extends StatefulWidget {
  /// 'signup' ouvre directement sur la création de compte ; sinon connexion.
  final String? initialMode;
  const AuthScreen({super.key, this.initialMode});

  @override
  State<AuthScreen> createState() => _AuthScreenState();
}

class _AuthScreenState extends State<AuthScreen> {
  final _formKey = GlobalKey<FormState>();
  final _emailCtrl = TextEditingController();
  final _passCtrl = TextEditingController();
  late bool _isLogin = widget.initialMode != 'signup';
  bool _loading = false;
  bool _resetSent = false;
  bool _obscurePass = true;
  String? _error;
  String? _versionLabel;
  late final String _headerPhoto =
      _authHeaderPhotos[Random().nextInt(_authHeaderPhotos.length)];

  @override
  void initState() {
    super.initState();
    _loadVersion();
  }

  Future<void> _loadVersion() async {
    final info = await PackageInfo.fromPlatform();
    if (mounted) {
      setState(() => _versionLabel = 'v${info.version} (${info.buildNumber})');
    }
  }

  @override
  void dispose() {
    _emailCtrl.dispose();
    _passCtrl.dispose();
    super.dispose();
  }

  /// `/onboarding/people` si l'utilisateur n'a encore aucune personne (donc
  /// à la création d'un compte), sinon `/home` directement — un compte
  /// existant qui se reconnecte a déjà passé cette étape.
  Future<String> _postAuthRoute() async {
    try {
      final tags = await TagService.myTags();
      final hasPerson =
          tags.any((t) => t.kind == 'personne' || t.kind == 'enfant');
      return hasPerson ? '/home' : '/onboarding/people';
    } catch (_) {
      return '/home';
    }
  }

  Future<void> _submit() async {
    if (!_formKey.currentState!.validate()) return;
    setState(() {
      _loading = true;
      _error = null;
    });
    try {
      if (_isLogin) {
        await FirebaseAuth.instance.signInWithEmailAndPassword(
            email: _emailCtrl.text.trim(), password: _passCtrl.text);
      } else {
        await FirebaseAuth.instance.createUserWithEmailAndPassword(
            email: _emailCtrl.text.trim(), password: _passCtrl.text);
      }
    } on FirebaseAuthException catch (e) {
      if (mounted) {
        setState(() {
          _error = _mapError(e.code);
          _loading = false;
        });
      }
      return;
    }
    try {
      await MigrationService.runIfNeeded();
    } catch (_) {}
    try {
      await UserService.onLogin();
    } catch (_) {}
    final route = await _postAuthRoute();
    if (!mounted) return;
    setState(() => _loading = false);
    context.go(route);
  }

  Future<void> _googleSignIn() async {
    setState(() {
      _loading = true;
      _error = null;
    });
    try {
      final gUser = await GoogleSignIn().signIn();
      if (gUser == null) {
        if (mounted) setState(() => _loading = false);
        return;
      }
      final gAuth = await gUser.authentication;
      await FirebaseAuth.instance.signInWithCredential(
          GoogleAuthProvider.credential(
              accessToken: gAuth.accessToken, idToken: gAuth.idToken));
    } catch (e) {
      // Cause la plus fréquente sur Android : aucune empreinte SHA-1 de
      // l'app enregistrée dans Firebase pour cette clé de signature
      // (google-services.json a alors un oauth_client vide côté Android, et
      // Google Sign-In échoue systématiquement avec ApiException 10). Loggé
      // en clair (pas juste avalé) pour pouvoir diagnostiquer sans device.
      debugPrint('Google Sign-In error: $e');
      if (mounted) {
        setState(() {
          _error = 'Erreur Google Sign-In';
          _loading = false;
        });
      }
      return;
    }
    try {
      await MigrationService.runIfNeeded();
    } catch (_) {}
    try {
      await UserService.onLogin();
    } catch (_) {}
    final route = await _postAuthRoute();
    if (!mounted) return;
    setState(() => _loading = false);
    context.go(route);
  }

  /// Passe par le backend (Resend) plutôt que
  /// `FirebaseAuth.sendPasswordResetEmail` : le relai email par défaut de
  /// Firebase Auth (domaine `*.firebaseapp.com` partagé entre des millions de
  /// projets) atterrit trop souvent en spam ou n'arrive jamais — signalé par
  /// David le 03.09.26. Le lien généré reste le vrai lien Firebase
  /// (`Admin SDK generatePasswordResetLink`, backend uniquement) ; seul
  /// l'envoi change. Pas de token Firebase à ce stade (utilisateur
  /// déconnecté) : appel HTTP direct, pas `BackendClient` (qui exige un
  /// utilisateur connecté).
  Future<void> _resetPassword() async {
    final email = _emailCtrl.text.trim();
    if (email.isEmpty || !email.contains('@')) {
      setState(() => _error = 'Saisis ton email d\'abord.');
      return;
    }
    setState(() {
      _loading = true;
      _error = null;
    });
    try {
      final response = await http
          .post(
            Uri.parse('${AppConfig.backendUrl}/api/notify/reset-password'),
            headers: {'Content-Type': 'application/json'},
            body: jsonEncode({'email': email}),
          )
          .timeout(const Duration(seconds: 20));
      if (mounted) {
        setState(() {
          _resetSent = response.statusCode == 200;
          _error = response.statusCode == 200
              ? null
              : 'Une erreur est survenue. Réessaie.';
          _loading = false;
        });
      }
    } catch (_) {
      if (mounted) {
        setState(() {
          _error = 'Une erreur est survenue. Réessaie.';
          _loading = false;
        });
      }
    }
  }

  String _mapError(String code) => switch (code) {
        'user-not-found' => 'Aucun compte avec cet email.',
        'wrong-password' ||
        'invalid-credential' =>
          'Email ou mot de passe incorrect.',
        'email-already-in-use' => 'Cet email est déjà utilisé.',
        'weak-password' => 'Mot de passe trop faible (6 caractères min).',
        _ => 'Une erreur est survenue. Réessaie.',
      };

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: AppColors.background,
      body: Column(
        children: [
          SizedBox(
            height: 220,
            width: double.infinity,
            child: Stack(
              fit: StackFit.expand,
              children: [
                CachedNetworkImage(
                  imageUrl: '$_headerPhoto?auto=format&fit=crop&w=900&q=70',
                  fit: BoxFit.cover,
                  placeholder: (_, __) => const ColoredBox(color: AppColors.surface),
                  errorWidget: (_, __, ___) => const ColoredBox(color: AppColors.surface),
                ),
                DecoratedBox(
                  decoration: BoxDecoration(
                    gradient: LinearGradient(
                      begin: Alignment.topCenter,
                      end: Alignment.bottomCenter,
                      colors: [
                        Colors.black.withOpacity(0.35),
                        Colors.black.withOpacity(0.15),
                        Colors.black.withOpacity(0.65),
                      ],
                      stops: const [0, 0.5, 1],
                    ),
                  ),
                ),
                Positioned(
                  top: 0,
                  left: 0,
                  child: SafeArea(
                    bottom: false,
                    child: Padding(
                      padding: const EdgeInsets.fromLTRB(14, 6, 0, 0),
                      child: GestureDetector(
                        onTap: () => context.canPop()
                            ? context.pop()
                            : context.go('/welcome'),
                        behavior: HitTestBehavior.opaque,
                        child: Container(
                          width: 34,
                          height: 34,
                          decoration: BoxDecoration(
                              color: Colors.white.withOpacity(0.18),
                              shape: BoxShape.circle),
                          child: const Icon(Icons.arrow_back,
                              color: Colors.white, size: 18),
                        ),
                      ),
                    ),
                  ),
                ),
                Positioned(
                  left: 24,
                  right: 24,
                  bottom: 18,
                  child: Row(
                    children: [
                      const Text('carnet',
                          style: TextStyle(
                            fontFamily: 'Fraunces',
                            fontStyle: FontStyle.italic,
                            fontSize: 26,
                            fontWeight: FontWeight.w600,
                            color: Colors.white,
                            height: 1,
                          )),
                      const Text('.',
                          style: TextStyle(
                            fontFamily: 'Fraunces',
                            fontStyle: FontStyle.italic,
                            fontSize: 26,
                            fontWeight: FontWeight.w700,
                            color: AppColors.sageDark,
                            height: 1,
                          )),
                    ],
                  ),
                ),
              ],
            ),
          ),
          Expanded(
            child: SafeArea(
              top: false,
              child: SingleChildScrollView(
                padding: const EdgeInsets.fromLTRB(28, 20, 28, 24),
                child: Form(
                  key: _formKey,
                  child: Column(
                    crossAxisAlignment: CrossAxisAlignment.start,
                    children: [
                      Text(_isLogin ? 'Bon retour' : 'Créer un compte',
                          style: const TextStyle(
                            fontFamily: 'Fraunces',
                            fontWeight: FontWeight.w600,
                            fontSize: 30,
                            color: AppColors.textDark,
                          )),
                const SizedBox(height: 6),
                Text(
                  _isLogin
                      ? 'Retrouve tes carnets.'
                      : 'Commence à capturer tes souvenirs.',
                  style: const TextStyle(color: AppColors.textMedium, fontSize: 14.5),
                ),
                const SizedBox(height: 32),
                TextFormField(
                  controller: _emailCtrl,
                  keyboardType: TextInputType.emailAddress,
                  validator: (v) =>
                      v == null || !v.contains('@') ? 'Email invalide' : null,
                  decoration: const InputDecoration(hintText: 'Email'),
                ),
                const SizedBox(height: 14),
                TextFormField(
                  controller: _passCtrl,
                  obscureText: _obscurePass,
                  validator: (v) =>
                      v == null || v.length < 6 ? '6 caractères minimum' : null,
                  decoration: InputDecoration(
                    hintText: 'Mot de passe',
                    suffixIcon: IconButton(
                      onPressed: () =>
                          setState(() => _obscurePass = !_obscurePass),
                      icon: Icon(
                        _obscurePass
                            ? Icons.visibility_off_outlined
                            : Icons.visibility_outlined,
                        color: AppColors.textMedium,
                        size: 20,
                      ),
                    ),
                  ),
                ),
                if (_error != null) ...[
                  const SizedBox(height: 12),
                  Row(children: [
                    const Icon(Icons.error_outline,
                        color: AppColors.error, size: 15),
                    const SizedBox(width: 6),
                    Expanded(
                        child: Text(_error!,
                            style: const TextStyle(
                                color: AppColors.error, fontSize: 12.5))),
                  ]),
                  // "Email ou mot de passe incorrect" ne distingue jamais
                  // (volontairement, anti-énumération) un vrai mauvais mot de
                  // passe d'un compte créé uniquement via Google, qui n'en a
                  // jamais eu — rappel discret plutôt que de laisser deviner
                  // (confusion vécue par David le 03.09.26).
                  if (_isLogin) ...[
                    const SizedBox(height: 6),
                    const Text(
                      'Connecté d\'habitude avec Google ? Utilise le bouton '
                      'ci-dessous.',
                      style:
                          TextStyle(color: AppColors.textMedium, fontSize: 12),
                    ),
                  ],
                ],
                if (_resetSent) ...[
                  const SizedBox(height: 12),
                  Row(children: const [
                    Icon(Icons.check_circle_outline,
                        color: AppColors.sageDark, size: 15),
                    SizedBox(width: 6),
                    Text('Email de réinitialisation envoyé.',
                        style: TextStyle(color: AppColors.sageDark, fontSize: 12.5)),
                  ]),
                ],
                const SizedBox(height: 22),
                SizedBox(
                  width: double.infinity,
                  child: _loading
                      ? const Center(
                          child: Padding(
                            padding: EdgeInsets.symmetric(vertical: 14),
                            child: CircularProgressIndicator(),
                          ),
                        )
                      : ElevatedButton(
                          onPressed: _submit,
                          child: Text(_isLogin ? 'Se connecter' : "S'inscrire"),
                        ),
                ),
                if (_isLogin)
                  Align(
                    alignment: Alignment.centerRight,
                    child: TextButton(
                      onPressed: _loading ? null : _resetPassword,
                      child: const Text('Mot de passe oublié ?',
                          style: TextStyle(fontSize: 13.5)),
                    ),
                  )
                else
                  const SizedBox(height: 8),
                Row(children: [
                  const Expanded(child: Divider(color: AppColors.border)),
                  Padding(
                    padding: const EdgeInsets.symmetric(horizontal: 14),
                    child: Text('ou',
                        style: TextStyle(color: AppColors.textMedium.withOpacity(0.8), fontSize: 12.5)),
                  ),
                  const Expanded(child: Divider(color: AppColors.border)),
                ]),
                const SizedBox(height: 14),
                SizedBox(
                  width: double.infinity,
                  child: OutlinedButton(
                    onPressed: _loading ? null : _googleSignIn,
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        Container(
                          width: 22,
                          height: 22,
                          decoration: const BoxDecoration(
                              color: Color(0xFF4285F4), shape: BoxShape.circle),
                          child: const Center(
                            child: Text('G',
                                style: TextStyle(
                                    color: Colors.white,
                                    fontWeight: FontWeight.bold,
                                    fontSize: 12)),
                          ),
                        ),
                        const SizedBox(width: 11),
                        const Text('Continuer avec Google'),
                      ],
                    ),
                  ),
                ),
                const SizedBox(height: 22),
                GestureDetector(
                  onTap: () => setState(() {
                    _isLogin = !_isLogin;
                    _error = null;
                    _resetSent = false;
                  }),
                  child: Center(
                    child: RichText(
                      text: TextSpan(
                        style: const TextStyle(
                            fontSize: 14, color: AppColors.textMedium),
                        children: [
                          TextSpan(
                              text: _isLogin
                                  ? 'Pas encore de compte ?  '
                                  : 'Déjà un compte ?  '),
                          TextSpan(
                            text: _isLogin ? 'S\'inscrire' : 'Se connecter',
                            style: const TextStyle(
                              color: AppColors.sageDark,
                              fontWeight: FontWeight.w700,
                              decoration: TextDecoration.underline,
                            ),
                          ),
                        ],
                      ),
                    ),
                  ),
                ),
                if (_versionLabel != null) ...[
                  const SizedBox(height: 22),
                  Center(
                    child: Text(_versionLabel!,
                        style: TextStyle(
                            fontSize: 11.5,
                            color: AppColors.textMedium.withOpacity(0.6))),
                  ),
                ],
                    ],
                  ),
                ),
              ),
            ),
          ),
        ],
      ),
    );
  }
}

File: kmp/composeApp/src/commonMain/kotlin/com/bloom/features/auth/AuthScreen.kt
package com.bloom.features.auth

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.text.KeyboardActions
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.text.input.PasswordVisualTransformation
import androidx.compose.ui.text.style.TextDecoration
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.bloom.core.di.ServiceLocator
import com.bloom.core.theme.AppColors
import com.bloom.features.navigation.AppRouter
import com.bloom.features.navigation.Screen
import kotlinx.coroutines.launch

@Composable
fun AuthScreen() {
    val scope = rememberCoroutineScope()
    var isLogin by remember { mutableStateOf(true) }
    var email by remember { mutableStateOf("") }
    var password by remember { mutableStateOf("") }
    var loading by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    var resetSent by remember { mutableStateOf(false) }

    fun submit() {
        if (email.isBlank() || !email.contains("@")) { error = "Email invalide"; return }
        if (password.length < 6) { error = "6 caractères minimum"; return }
        scope.launch {
            loading = true; error = null
            val result = if (isLogin)
                ServiceLocator.authService.signIn(email, password)
            else
                ServiceLocator.authService.signUp(email, password)
            result.onSuccess { AppRouter.navigateAndClear(Screen.Home) }
                .onFailure { error = mapError(it.message ?: "") }
            loading = false
        }
    }

    fun sendReset() {
        if (email.isBlank() || !email.contains("@")) { error = "Saisis ton email d'abord."; return }
        scope.launch {
            loading = true
            ServiceLocator.authService.sendPasswordReset(email)
                .onSuccess { resetSent = true }
                .onFailure { error = "Une erreur est survenue." }
            loading = false
        }
    }

    Box(
        modifier = Modifier.fillMaxSize().background(AppColors.Beige),
        contentAlignment = Alignment.Center,
    ) {
        Column(
            modifier = Modifier
                .widthIn(max = 480.dp)
                .fillMaxWidth()
                .verticalScroll(rememberScrollState())
                .padding(24.dp),
        ) {
            Spacer(Modifier.height(40.dp))
            Text("🌸", fontSize = 64.sp)
            Spacer(Modifier.height(16.dp))
            Text(
                text = if (isLogin) "Bon retour !" else "Créer un compte",
                style = MaterialTheme.typography.headlineLarge.copy(fontWeight = FontWeight.Bold),
            )
            Spacer(Modifier.height(8.dp))
            Text(
                text = if (isLogin) "Retrouve les souvenirs de ton enfant."
                else "Commence à capturer les premiers instants.",
                color = AppColors.TextMedium,
            )
            Spacer(Modifier.height(32.dp))

            OutlinedTextField(
                value = email,
                onValueChange = { email = it; error = null },
                label = { Text("Email") },
                modifier = Modifier.fillMaxWidth(),
                keyboardOptions = KeyboardOptions(
                    keyboardType = KeyboardType.Email,
                    imeAction = ImeAction.Next,
                ),
                singleLine = true,
                colors = bloomTextFieldColors(),
            )
            Spacer(Modifier.height(12.dp))
            OutlinedTextField(
                value = password,
                onValueChange = { password = it; error = null },
                label = { Text("Mot de passe") },
                modifier = Modifier.fillMaxWidth(),
                visualTransformation = PasswordVisualTransformation(),
                keyboardOptions = KeyboardOptions(
                    keyboardType = KeyboardType.Password,
                    imeAction = ImeAction.Done,
                ),
                keyboardActions = KeyboardActions(onDone = { submit() }),
                singleLine = true,
                colors = bloomTextFieldColors(),
            )

            error?.let {
                Spacer(Modifier.height(8.dp))
                Text(it, color = AppColors.Error, style = MaterialTheme.typography.bodySmall)
            }
            if (resetSent) {
                Spacer(Modifier.height(8.dp))
                Card(colors = CardDefaults.cardColors(containerColor = AppColors.Sage.copy(alpha = 0.1f))) {
                    Row(Modifier.padding(12.dp), verticalAlignment = Alignment.CenterVertically) {
                        Text("✓", color = AppColors.Sage)
                        Spacer(Modifier.width(8.dp))
                        Text("Email envoyé ! Vérifie ta boîte mail.", color = AppColors.Sage, fontSize = 13.sp)
                    }
                }
            }
            Spacer(Modifier.height(20.dp))

            if (loading) {
                Box(Modifier.fillMaxWidth(), contentAlignment = Alignment.Center) {
                    CircularProgressIndicator(color = AppColors.Sage)
                }
            } else {
                Button(
                    onClick = { submit() },
                    modifier = Modifier.fillMaxWidth().height(52.dp),
                    colors = ButtonDefaults.buttonColors(containerColor = AppColors.Sage),
                    shape = MaterialTheme.shapes.medium,
                ) {
                    Text(if (isLogin) "Se connecter" else "S'inscrire", fontWeight = FontWeight.Medium)
                }
            }

            if (isLogin) {
                TextButton(
                    onClick = { sendReset() },
                    enabled = !loading,
                    modifier = Modifier.fillMaxWidth(),
                ) {
                    Text("Mot de passe oublié ?", color = AppColors.Sage, fontSize = 13.sp)
                }
            }

            Spacer(Modifier.height(16.dp))
            Divider(color = AppColors.SoftGray.copy(alpha = 0.3f))
            Spacer(Modifier.height(16.dp))

            Text(
                text = if (isLogin) "Pas encore de compte ? S'inscrire"
                else "Déjà un compte ? Se connecter",
                color = AppColors.Sage,
                style = MaterialTheme.typography.bodyMedium.copy(
                    textDecoration = TextDecoration.Underline,
                ),
                modifier = Modifier
                    .fillMaxWidth()
                    .clickable {
                        isLogin = !isLogin; error = null; resetSent = false
                    }
                    .padding(8.dp),
            )
        }
    }
}

@Composable
private fun bloomTextFieldColors() = OutlinedTextFieldDefaults.colors(
    focusedBorderColor = AppColors.Sage,
    unfocusedBorderColor = AppColors.SoftGray,
    focusedLabelColor = AppColors.Sage,
    cursorColor = AppColors.Sage,
)

private fun mapError(msg: String): String = when {
    "EMAIL_NOT_FOUND" in msg || "user-not-found" in msg -> "Aucun compte avec cet email."
    "INVALID_PASSWORD" in msg || "wrong-password" in msg -> "Mot de passe incorrect."
    "EMAIL_EXISTS" in msg || "email-already-in-use" in msg -> "Cet email est déjà utilisé."
    "WEAK_PASSWORD" in msg || "weak-password" in msg -> "Mot de passe trop faible (6 caractères min)."
    else -> "Une erreur est survenue. Réessaie."
}
